x-pack/filebeat/input/httpjson/request.go

```diff
@@ -424,11 +471,190 @@ func (r *requester) processAndPublishEvents(stdCtx context.Context, trCtx *trans
 		}
 		trCtx.updateLastEvent(maybeMsg.msg)
 		trCtx.updateCursor()
+
 		n++
 	}
 	return n
 }
 
+// processAndPublishChainEvents processes and publishes chain events based on response type,
+// this separate func has been introduced , since each chain step now has its own response processor and paginator
+func processAndPublishChainEvents(stdCtx context.Context, trCtx *transformContext, chainRsp *responseProcessor, publisher inputcursor.Publisher, finalResps []*http.Response, publish bool, log *logp.Logger) int {
```

Comment:
The body of this is almost identical to `*requester.processAndPublishEvents`. The common code should be factored out. I would say that there is only a need for one, which takes `events` and `log`. In the case that currently calls `*requester.processAndPublishEvents` then becomes `shinyNewProcessAndPublishEvents(stdCtx, chainTrCtx, events, rf.chainResponseProcessor, publisher, resps, i < len(r.requestFactories), r.log)`, where `events` is made depending on context.

Reply:
have fixed this

Reply:
I'd place the events parameter earlier in the signature since it's the focus of the function (only after the contexts because of the context first convention). To avoid the issue of long code inserted into the caller assign to a var and use that.